experimental/hip/native_executable.c

```diff
@@ -0,0 +1,309 @@
+                            "flatbuffer data is not present");
+  }
+
+  // Run flatcc generated verification. This ensures all pointers are in-bounds
+  // and that we can safely walk the file, but not that the actual contents of
+  // the flatbuffer meet our expectations.
+  int verify_ret = iree_hal_hip_ExecutableDef_verify_as_root(
+      flatbuffer_data.data, flatbuffer_data.data_length);
+  if (verify_ret != flatcc_verify_ok) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "flatbuffer verification failed: %s",
+                            flatcc_verify_error_string(verify_ret));
+  }
+
+  iree_hal_hip_ExecutableDef_table_t executable_def =
+      iree_hal_hip_ExecutableDef_as_root(flatbuffer_data.data);
+
+  flatbuffers_string_vec_t entry_points_vec =
+      iree_hal_hip_ExecutableDef_entry_points_get(executable_def);
+  size_t entry_point_count = flatbuffers_string_vec_len(entry_points_vec);
+  for (size_t i = 0; i < entry_point_count; ++i) {
+    if (flatbuffers_string_len(
+            flatbuffers_string_vec_at(entry_points_vec, i)) == 0) {
+      return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "executable entry point %zu has no name", i);
+    }
+  }
+
+  iree_hal_hip_BlockSizeDef_vec_t block_sizes_vec =
+      iree_hal_hip_ExecutableDef_block_sizes_get(executable_def);
+  size_t block_size_count = iree_hal_hip_BlockSizeDef_vec_len(block_sizes_vec);
+  if (block_size_count == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "no block sizes present");
+  }
+
+  if (entry_point_count != block_size_count) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "entry points (%zu) and block sizes (%zu) count mismatch",
+        entry_point_count, block_size_count);
+  }
+
+  flatbuffers_string_t hsaco_image =
+      iree_hal_hip_ExecutableDef_hsaco_image_get(executable_def);
+  if (flatbuffers_string_len(hsaco_image) == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "no HSACO image present");
+  }
+
+  return iree_ok_status();
+}
+
+iree_status_t iree_hal_hip_native_executable_create(
+    const iree_hal_hip_dynamic_symbols_t* symbols, hipDevice_t device,
+    const iree_hal_executable_params_t* executable_params,
+    iree_allocator_t host_allocator, iree_hal_executable_t** out_executable) {
+  IREE_ASSERT_ARGUMENT(executable_params);
+  IREE_ASSERT_ARGUMENT(out_executable);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  *out_executable = NULL;
+  iree_hal_hip_native_executable_t* executable = NULL;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_hip_native_executable_flatbuffer_verify(
+              executable_params->executable_data));
+
+  iree_hal_hip_ExecutableDef_table_t executable_def =
+      iree_hal_hip_ExecutableDef_as_root(
+          executable_params->executable_data.data);
+
+  flatbuffers_string_t hsaco_image =
```

Comment:
Nit: can we order these to follow the definition order in the schema. It's more consistent that way.

Feel free to update cuda2 side too.